hathor/transaction/base_transaction.py

```diff
@@ -424,8 +425,81 @@ def get_struct(self) -> bytes:
         struct_bytes += self.get_struct_nonce()
         return struct_bytes
 
+    def get_all_dependencies(self) -> Set[bytes]:
+        """Set of all tx-hashes needed to fully validate this tx, including parent blocks/txs and inputs."""
+        from hathor.transaction.genesis import GENESIS_HASHES
+        return set(chain(self.parents, (i.tx_id for i in self.inputs))) - set(GENESIS_HASHES)
+
+    def get_tx_dependencies(self) -> Set[bytes]:
+        """Set of all tx-hashes needed to fully validate this, except for block parent, i.e. only tx parents/inputs."""
+        from hathor.transaction.genesis import GENESIS_HASHES
+        parents = self.parents[1:] if self.is_block else self.parents
+        return set(chain(parents, (i.tx_id for i in self.inputs))) - set(GENESIS_HASHES)
+
+    def get_tx_parents(self) -> Set[bytes]:
+        """Set of parent tx hashes, typically used for syncing transactions."""
+        from hathor.transaction.genesis import GENESIS_HASHES
+        return set(self.parents[1:] if self.is_block else self.parents) - set(GENESIS_HASHES)
+
+    def can_validate_full(self) -> bool:
+        """ Check if this transaction is ready to be fully validated, either all deps are full-valid or one is invalid.
+        """
+        assert self.storage is not None
+        deps = self.get_all_dependencies()
+        return (
```

Comment:
Using for loops is annoying because it adds a lot of breaks to the flow, I could pull an updated version of this implementation that breaks this up into more methods which makes it much better to use for loops.